efficientdet/keras/train.py

```diff
@@ -0,0 +1,266 @@
+                     'Number of examples in one epoch')
+flags.DEFINE_integer('num_epochs', None, 'Number of epochs for training')
+flags.DEFINE_string('mode', 'train',
+                    'Mode to run: train or eval (default: train)')
+flags.DEFINE_string('model_name', 'efficientdet-d1',
+                    'Model name: retinanet or efficientdet')
+flags.DEFINE_bool('eval_after_training', False, 'Run one eval after the '
+                  'training finishes.')
+flags.DEFINE_bool('debug', False, 'Debug mode')
+
+# For Eval mode
+flags.DEFINE_integer('min_eval_interval', 180,
+                     'Minimum seconds between evaluations.')
+flags.DEFINE_integer(
+    'eval_timeout', None,
+    'Maximum seconds between checkpoints before evaluation terminates.')
+
+FLAGS = flags.FLAGS
+
+
+def main(_):
+  # Parse and override hparams
+  config = hparams_config.get_detection_config(FLAGS.model_name)
+  config.override(FLAGS.hparams)
+  if FLAGS.num_epochs:  # NOTE: remove this flag after updating all docs.
+    config.num_epochs = FLAGS.num_epochs
+
+  # Parse image size in case it is in string format.
+  config.image_size = utils.parse_image_size(config.image_size)
+
+  if FLAGS.use_xla and FLAGS.strategy != 'tpu':
+    tf.config.optimizer.set_jit(True)
+    tf.config.experimental.set_memory_growth(True)
+
+  if FLAGS.debug:
+    tf.config.experimental_run_functions_eagerly(True)
+    tf.debugging.set_log_device_placement(True)
+    tf.random.set_seed(111111)
+    logging.set_verbosity(logging.DEBUG)
+
+  if FLAGS.strategy == 'tpu':
+    tpu_cluster_resolver = tf.distribute.cluster_resolver.TPUClusterResolver(
+        FLAGS.tpu, zone=FLAGS.tpu_zone, project=FLAGS.gcp_project)
+    tf.config.experimental_connect_to_cluster(tpu_cluster_resolver)
+    tf.tpu.experimental.initialize_tpu_system(tpu_cluster_resolver)
+    strategy = tf.distribute.experimental.TPUStrategy(tpu_cluster_resolver)
+
+    logging.info("All devices: ", tf.config.list_logical_devices('TPU'))
+  elif FLAGS.strategy == 'gpus':
+    strategy = tf.distribute.MirroredStrategy()
+    per_replica_train_batch_size = FLAGS.train_batch_size // strategy.num_replicas_in_sync
+    per_replica_eval_batch_size = FLAGS.eval_batch_size // strategy.num_replicas_in_sync
+    logging.info("All devices: ", tf.config.list_physical_devices('GPU'))
+  else:
+    strategy = None
+    per_replica_train_batch_size = FLAGS.train_batch_size
+    per_replica_eval_batch_size = FLAGS.eval_batch_size
+
+  # Check data path
+  if FLAGS.mode in ('train',
+                    'train_and_eval') and FLAGS.training_file_pattern is None:
+    raise RuntimeError('You must specify --training_file_pattern for training.')
+  if FLAGS.mode in ('eval', 'train_and_eval'):
+    if FLAGS.validation_file_pattern is None:
+      raise RuntimeError('You must specify --validation_file_pattern '
+                         'for evaluation.')
+
+  # The following is for spatial partitioning. `features` has one tensor while
```

Comment:
This spatial partition code is optional: it is a legacy code, and I never use it for training.
Feel free to delete them if that helps reduce the coding complexity.